### PR TITLE
Support range/length refinement providers for enums

### LINKED ISSUE #1282: Invalid code generated when constraints on enums are used
```smithy
$version: "2"

namespace hello

structure MyStruct {
    @length(max: 3)
    s: Str
}

enum Str {
    FOO
    BARS
}
```

This fails to compile on 0.18.2:

```
[error] /Users/kubukoz/projects/demos/target/scala-2.13/src_managed/main/scala/hello/MyStruct.scala:17:25: could not find implicit value for parameter constraint: smithy4s.RefinementProvider.Simple[smithy.api.Length,hello.Str]
[error]     Str.schema.validated(smithy.api.Length(min = None, max = Some(3L))).optional[MyStruct]("s", _.s),
[error]                         ^
[error] /Users/kubukoz/projects/demos/target/scala-2.13/src_managed/main/scala/hello/MyStruct.scala:19:14: missing argument list for method apply in object MyStruct
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `apply _` or `apply(_)` instead of `apply`.
[error]     MyStruct.apply
[error]              ^
[error] two errors found
[error] (Compile / compileIncremental) Compilation failed
```

because there's no `RefinementProvider` available for enums, or a bijection that would allow deriving one.

BTW. `@length` / `@range` on the enum/intEnum shape itself are ignored by the codegen, including `val hints`.

Technically, the Smithy validators allow `@length` on enums, and `@range` on int enums. Should we support this somehow? I'm not sure there's a usecase in limiting the amount of allowed enum values further by imposing constraints, but I also can't say that nobody is doing so.

### PR DESCRIPTION
**Pull request by Baccata**
Fixes https://github.com/disneystreaming/smithy4s/issues/1282

## PR Checklist (not all items are relevant to all PRs)

- [x] Added bootstrapped code 
- [x] Updated changelog

### CODE REVIEW

File: modules/core/src/smithy4s/RefinementProvider.scala [hunk 1]
```diff
@@ -82,6 +83,12 @@ object RefinementProvider {
   implicit def mapLengthConstraint[K, V]: Simple[Length, Map[K, V]] =
     new LengthConstraint[Map[K, V]](_.size)
 
+  implicit def enumLengthConstraint[E <: Enumeration.Value]: Simple[Length, E] =
```
Comment (kubukoz) [inline modules/core/src/smithy4s/RefinementProvider.scala]: **question:** should we also have `@pattern`?

Comment (kubukoz) [reply modules/core/src/smithy4s/RefinementProvider.scala]: `@idRef` would be nasty. If someone ever creates an enum that matches the ShapeId format, I'll personally go to their house and remove their github account.

Comment (Baccata) [reply modules/core/src/smithy4s/RefinementProvider.scala]: > question: should we also have @pattern? 

Is it allowed ? you mentioned range and length, but ...

Comment (kubukoz) [reply modules/core/src/smithy4s/RefinementProvider.scala]: I didn't check the others! 😅

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # 0.18.3
 
+* Support length/range constraints on structure fields targeting enums
```
Comment (kubukoz) [inline CHANGELOG.md]: nitpick: 

```suggestion
* Support constraint traits on members targeting enums
```

**Review by kubukoz — APPROVED** [review]

**Pull request merged**